src/txreconciliation.cpp

```diff
@@ -0,0 +1,893 @@
+     * (currently, transactions received from inbound links)
+     * to some of the peers:
+     * - all pre-reconciliation peers supporting transaction relay;
+     * - a limited number of outbound reconciling peers *for which this flag is enabled*.
+     * We enable this flag based on whether we have a
+     * sufficient number of outbound transaction relay peers.
+     * This flooding makes transaction relay across the network faster
+     * without introducing high the bandwidth overhead.
+     * Transactions announced via flooding should not be added to
+     * the reconciliation set.
+     */
+    bool m_flood_to;
+
+    /**
+     * Reconciliation involves computing and transmitting sketches,
+     * which is a bandwidth-efficient representation of transaction IDs.
+     * Since computing sketches over full txID is too CPU-expensive,
+     * they will be computed over shortened IDs instead.
+     * These short IDs will be salted so that they are not the same
+     * across all pairs of peers, because otherwise it would enable network-wide
+     * collisions which may (intentionally or not) halt relay of certain transactions.
+     * Both of the peers contribute to the salt.
+     */
+    const uint64_t m_k0, m_k1;
+
+    /**
+     * Store all transactions which we would relay to the peer (policy checks passed, etc.)
+     * in this set instead of announcing them right away. When reconciliation time comes, we will
+     * compute an efficient representation of this set ("sketch") and use it to efficient reconcile
+     * this set with a similar set on the other side of the connection.
+     */
+    std::set<uint256> m_local_set;
+
+    /**
+     * A reconciliation round may involve an extension, which is an extra exchange of messages.
+     * Since it may happen after a delay (at least network latency), new transactions may come
+     * during that time. To avoid mixing old and new transactions, those which are subject for
+     * extension of a current reconciliation round are moved to a reconciliation set snapshot
+     * after an initial (non-extended) sketch is sent.
+     * New transactions are kept in the regular reconciliation set.
+     */
+    std::set<uint256> m_local_set_snapshot;
+
+    /**
+     * A reconciliation round may involve an extension, in which case we should remember
+     * a capacity of the sketch sent out initially, so that a sketch extension is of the same size.
+     */
+    uint16_t m_capacity_snapshot{0};
+
+    /**
+     * Reconciliation sketches are computed over short transaction IDs.
+     * This is a cache of these IDs enabling faster lookups of full wtxids,
+     * useful when peer will ask for missing transactions by short IDs
+     * at the end of a reconciliation round.
+     */
+    std::map<uint32_t, uint256> m_local_short_id_mapping;
+
+    /** Keep track of reconciliations with the peer. */
+    ReconciliationInitByUs m_state_init_by_us;
+    ReconciliationInitByThem m_state_init_by_them;
```

Comment:
What purpose serve those 2 different structures ? Can't they be aggregated in one ?

Reply:
They have very different internals, and I think it's for ease of understanding. If we are initiator, we don't care about those fields in other struct.

Both of them could have been inheritance child classes of ReconciliationState (but with extra fields), but I think in C++ specifically it's easier to implement them as extension classes like this.

Reply:
Okay I took the point around ease of understanding. Maybe you can document around `m_state_init_by_us`/`m_state_init_by_them` in `ReconciliationState` why in protocol flow only one of them matters.